### PR TITLE
CXX-3194 Migrate from macos-1100 to macos-14-arm64

### PR DESCRIPTION
**Pull request by eramongodb**
Similar to https://github.com/mongodb/mongo-c-driver/pull/1791, motivated by [DEVPROD-21](https://jira.mongodb.org/browse/DEVPROD-21). Verified by [this patch](https://spruce.mongodb.com/version/674f3b6232d276000785a7ce/).

MongoDB Server 5.0 tasks remain on `macos-14` (Intel) due to lack of availability on `macos-14-arm64`.

Also an attempt to avoid spurious Clang internal compiler errors as observed in [this patch](https://spruce.mongodb.com/task/mongo_cxx_driver_integration_matrix_integration_macos_1100_release_static_extra_alignment_latest_single_e55829024adbd437e48d8310bf959f644d67dc07_24_12_02_21_11_39/).

```
clang: error: unable to execute command: Segmentation fault: 11
clang: error: clang frontend command failed due to signal (use -v to see invocation)
Apple clang version 12.0.0 (clang-1200.0.32.29)
Target: x86_64-apple-darwin20.6.0
Thread model: posix
```

The addition of `-Wno-aligned-mismatch` was necessary to address compilation errors similar to those currently being silenced by `-Wno-aligned-new` for GCC (see "Extra Alignment" in PR description of https://github.com/mongodb/mongo-cxx-driver/pull/1244).

### CODE REVIEW

File: .evergreen/config_generator/components/integration.py [hunk 1]
```diff
@@ -34,7 +34,8 @@
     ('debian12', None, ['Release'], ['shared',         ], [      20], [None], ['plain',        ], [False, True], ['latest'], ['single',                     ]),
     ('debian12', None, ['Release'], ['shared',         ], [None,   ], [None], [         'csfle'], [False,     ], ['latest'], [          'replica', 'sharded']),
 
-    ('macos-1100', None, ['Release'], ['shared', 'static'], [None], [None], ['plain'], [False, True], ['5.0', 'latest'], ['single']),
+    ('macos-14',       None, ['Release'], ['shared', 'static'], [None], [None], ['plain'], [False, True], ['5.0',         ], ['single']),
```
Comment (kevinAlbs) [inline .evergreen/config_generator/components/integration.py:37]: I am unsure of the user impact for dropping support of macOS 11, 12, and 13. macOS 13 is [still supported](https://endoflife.date/macos). I filed CXX-3193 to request dropping macOS 11 and 12.

Suggest documenting support for macOS 11 and 12 is deprecated in CHANGELOG.md to give users a heads up. Restore one macOS 11 task to ensure the driver continues to build on macOS 11.

**Review by kevinAlbs — CHANGES_REQUESTED** [review]

File: .evergreen/config_generator/components/integration.py (hunk 1 above)
Comment (eramongodb) [reply .evergreen/config_generator/components/integration.py:37]: Jumping from MacOS 11 to MacOS 14 distros is motivated by [DEVPROD-8869](https://jira.mongodb.org/browse/DEVPROD-8869) and [DEVPROD-8870](https://jira.mongodb.org/browse/DEVPROD-8870). I do not think our test coverage needs to be strictly 1-to-1 with platforms we (do not) document as actively supported. LMK if we nevertheless want to maintain coverage of MacOS 11 (+ perhaps 13; there is no distro for 12) until we are forced to move away once the distros are decommissioned.

Comment (eramongodb) [reply .evergreen/config_generator/components/integration.py:39]: Restored some MacOS 11 coverage (integration matrix, latest only) after further discussion. Included a deprecation notice in the changelog as suggested. An attempt to migrate from `macos-11` to `macos-11-arm64` (as done in https://github.com/mongodb/mongo-c-driver/pull/1791) was blocked by [DEVPROD-13222](https://jira.mongodb.org/browse/DEVPROD-13222).

**Review by kevinAlbs — APPROVED** [review]

**Comment by eramongodb** [comment]
The `macos-11` distro appears to have observable differences from the `macos-1100` distro concerning the presence of a thread library as (not) detected by CMake's FindThreads module. Reverted to `macos-1100` for now; a followup PR will complete the transition to `macos-11-arm64` once [DEVPROD-13222](https://jira.mongodb.org/browse/DEVPROD-13222) is resolved.

**Pull request merged**